### PR TITLE
gh-117398: Add datetime C-API type check test for subinterpreters

### PR DESCRIPTION
**Pull request by neonene**
Check if the `DateType` C-API type matches the `datetime.date` type on main and shared/isolated subinterpreters.

@ericsnowcurrently  Please review if this usage is not invalid.

* Issue: gh-117398

### CODE REVIEW

File: Modules/_testmultiphase.c [hunk 1]
```diff
@@ -969,3 +969,43 @@ PyInit__test_shared_gil_only(void)
...
+};
+
+static PyModuleDef datetime_capi_client_def = TEST_MODULE_DEF(
+    "_testmultiphase_datetime_capi_client", datetime_capi_client_slots, datetime_capi_client_methods);
+
+PyMODINIT_FUNC
+PyInit__test_datetime_capi_client(void)
+{
+    return PyModuleDef_Init(&datetime_capi_client_def);
+}
```
Comment (ericsnowcurrently) [inline Modules/_testmultiphase.c]: _testmultiphase is meant to exercise the capabilities of the import machinery, so it seems like an odd place for this.  Why not add the method to Modules/_testcapi/datetime.c?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
The test makes sense.  Thanks for adding it.  I have one question.

**Comment by neonene** [comment]
I have made the requested changes; please review again.

**Comment by ericsnowcurrently** [comment]
It looks like part of your PR was removed by mistake.  I thought the new test in `test_capi.test_misc` would still be there.

**Comment by neonene** [comment]
Restored and reconsidered the test for subinterpreters.

File: Lib/test/test_capi/test_misc.py [hunk 2]
```diff
@@ -2285,6 +2285,46 @@ def test_module_state_shared_in_global(self):
         subinterp_attr_id = os.read(r, 100)
         self.assertEqual(main_attr_id, subinterp_attr_id)
 
+    @unittest.skipIf(_testmultiphase is None, "test requires _testmultiphase module")
```
Comment (ericsnowcurrently) [inline Lib/test/test_capi/test_misc.py]: This isn't needed any more, right?

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

Thanks for the extra effort on getting the test right.  I really like what you've done.

There's one minor thing I noticed but otherwise I think this is good to go.

File: Lib/test/test_capi/test_misc.py (hunk 2 above)
Comment (neonene) [reply Lib/test/test_capi/test_misc.py]: Right. I removed it and tried clearer way to skip the test on WASI: b89801c679f44d4d75e32bd4713567183a6ae58f

File: Lib/test/test_capi/test_misc.py [hunk 3]
```diff
@@ -2285,6 +2285,52 @@ def test_module_state_shared_in_global(self):
...
+            exec(script)
+        with self.subTest("non-isolated subinterpreter"):
+            ret = support.run_in_subinterp(script)
+            self.assertEqual(ret, 0)
+        if _interpreters:
+            with self.subTest("isolated subinterpreter"):
+                interpid = _interpreters.create()
+                ret = _interpreters.run_string(interpid, script)
+                _interpreters.destroy(interpid)
+                self.assertIsNone(ret)
```
Comment (ericsnowcurrently) [inline Lib/test/test_capi/test_misc.py]: (nothing do do here)

FYI, you can also use `support.run_in_subinterp_with_config()`, instead of using the `_interpreters` module.  However, what you have here is also good.

File: Lib/test/test_capi/test_misc.py [hunk 4]
```diff
@@ -2285,6 +2285,52 @@ def test_module_state_shared_in_global(self):
...
+                run(module.check_datetime, _datetime.datetime.now())
+                run(module.check_time,     _datetime.time(12, 30))
+                run(module.check_delta,    _datetime.timedelta(1))
+                run(module.check_tzinfo,   _datetime.tzinfo())
+        """)
+        with self.subTest("main interpreter"):
+            exec(script)
+        with self.subTest("non-isolated subinterpreter"):
+            ret = support.run_in_subinterp(script)
+            self.assertEqual(ret, 0)
```
Comment (ericsnowcurrently) [inline Lib/test/test_capi/test_misc.py]: (nothing to do here)

FWIW, in my last review I was going to leave a comment suggesting that you move this subtest under the `if _interpreters:` block, even though the feature is still available when the module isn't, but for consistency.  However, I realized it's worth having this subtest always run.  I'm mentioning this here (belatedly) because I thought it might be a helpful observation to provide some extra context about the feature.  Nothing needs to change though.

**Review by ericsnowcurrently — APPROVED** [review]
still LGTM

I've left some extra commentary that needs zero changes.  Thanks again for working on this!

**Comment by neonene** [comment]
I'll try something more.

**Comment by neonene** [comment]
Done. It would be nice if the WASI CI could load the test module without depending on `ExtensionFileLoader`.

**Comment by neonene** [comment]
Reconsidered.  The `_interpreters` module is mostly available unless `_testcapi` is a built-in module, right?

**Comment by neonene** [comment]
This test now requires `_datetime` with multi-phase init.

**Comment by neonene** [comment]
Please review again.

File: Lib/test/test_capi/test_misc.py [hunk 5]
```diff
@@ -2263,6 +2263,44 @@ def test_module_state_shared_in_global(self):
         subinterp_attr_id = os.read(r, 100)
         self.assertEqual(main_attr_id, subinterp_attr_id)
 
+    def test_datetime_capi_type_check(self):
```
Comment (ericsnowcurrently) [inline Lib/test/test_capi/test_misc.py]: All our other tests of the datetime C-API are methods of `CapiTest` in Lib/test/datetimetester.py.  Is there any reason this test should be different?

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Mostly LGTM

The test looks great.  The only change I'd recommend is to move it to Lib/test/datetimetester.py.

**Comment by neonene** [comment]
Moved.  I've touched `support.run_in_subinterp_with_config()` as well.

I have made the requested changes; please review again.

File: Lib/test/support/__init__.py [hunk 6]
```diff
@@ -1801,14 +1801,11 @@ def run_in_subinterp_with_config(code, *, own_gil=None, **config):
...
-        if gil == 0:
-            config['gil'] = 'default'
-        elif gil == 1:
-            config['gil'] = 'shared'
-        elif gil == 2:
-            config['gil'] = 'own'
-        else:
+        if gil in gil_supported:
+            config['gil'] = gil_supported[gil]
+        elif gil not in gil_supported.values():
```
Comment (ericsnowcurrently) [inline Lib/test/support/__init__.py]: Hmm, new code is a little harder to follow.  What's the motivation for this change?  Is it only that we aren't correctly handling the case where `config['gil']` is already a string?  If so, I'd rather we keep things simple.

The simplest thing is to allow the value to be a string:

```python
    else:
        gil = config['gil']
        if gil == 0:
            config['gil'] = 'default'
        elif gil == 1:
            config['gil'] = 'shared'
        elif gil == 2:
            config['gil'] = 'own'
        elif not isinstance(gil, str):
            raise NotImplementedError(gil)
```

That's a 1-line change and benefits readers.

Comment (ericsnowcurrently) [reply Lib/test/support/__init__.py]: FWIW, the ideal solution would be to update `interp_config_from_dict()` (in Python/interpconfig.c) to support passing an int for `config['gil']`.  Then the else block here could go away.

While we're at it we could also update `_init_interp_config_from_object()` (in Modules/_testinternalcapi.c) to let "config" be a dict.  Then we don't need to create an otherwise unnecessary `types.SimpleNamespace` here.

All that should be handled separately from this PR though.

**Review by ericsnowcurrently — CHANGES_REQUESTED** [review]
Thanks for moving that test!  There's just one thing to address and then this PR should be ready to merge.  Thanks for your patience.

File: Lib/test/support/__init__.py (hunk 6 above)
Comment (neonene) [reply Lib/test/support/__init__.py]: I just noticed that `_interpreters.new_config()` emits `ValueError` when the gil name has a typo.  Maybe this suffices?

**Comment by neonene** [comment]
I have made the requested changes; please review again.

File: Lib/test/datetimetester.py [hunk 7]
```diff
@@ -6780,6 +6785,42 @@ def test_datetime_from_timestamp(self):
 
                     self.assertEqual(dt_orig, dt_rt)
 
+    def test_type_check_in_subinterp(self):
+        script = textwrap.dedent(f"""
+            if {_interpreters is None}:
```
Comment (ericsnowcurrently) [inline Lib/test/datetimetester.py:6790]: Ah, somehow my brain was ignoring the curly braces (maybe thought parentheses?) and I was wondering how `_interpreters` was already bound in the subinterpreter. LOL

**Review by ericsnowcurrently — APPROVED** [review]
LGTM

File: Lib/test/datetimetester.py (hunk 7 above)
Comment (neonene) [reply Lib/test/datetimetester.py:6790]: I agree that this may be confusing for readers.

Comment (ericsnowcurrently) [reply Lib/test/datetimetester.py:6790]: Nah, it's fine.  It's mostly an artifact of my working in C a lot lately and my font being really, really small. 😄
